### PR TITLE
Add SoftDot node

### PR DESCRIPTION
**Pull request by abpolym**
This PR is regarding to issue #231 .

### CODE REVIEW

File: src/nodes/softdot.jl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import StatsFuns: log2π
+
+struct SoftDot end
+const softdot = SoftDot
+
+@node softdot Stochastic [y, (θ, aliases = [theta]), x, (γ, aliases = [gamma])]
+
+# TODO: Generalize q_γ for GammaDistributionsFamily by comparing derivations for GammaShapeRate
+@average_energy softdot (q_y::NormalDistributionsFamily, q_θ::NormalDistributionsFamily, q_x::NormalDistributionsFamily, q_γ::GammaShapeRate) = begin
```
Comment (albertpod) [inline src/nodes/softdot.jl:11]: You can safely use GammaDistributionFamily here.

I would suggest to remove multiline AE calculations and have smth like:

```julia
function average_energy(q_y::NormalDistributionsFamily, q_θ::NormalDistributionsFamily, q_x::NormalDistributionsFamily, q_γ::GammaDistributionsFamily)
    m_y, V_y = mean_cov(q_y)
    m_θ, V_θ = mean_cov(q_θ)
    m_x, V_x = mean_cov(q_x)
    m_γ = mean(q_γ)
    AE = -mean(log.(q_γ)) + log2π .+ m_γ .* (V_y + m_y.^2 .- 2m_γ * (m_θ'*m_x) .+ mul_trace(V_θ, V_x) .+ m_x' * V_θ * m_x .+ m_θ' * (V_x .+ m_x * m_x') * m_θ) ./ 2
    return AE
end
```

File: src/nodes/softdot.jl [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    m_θ, V_θ = mean_cov(q_θ)
+    m_x, V_x = mean_cov(q_x)
+    m_γ = mean(q_γ)
+    AE = 0
+    AE -= mean(log, q_γ)
+    AE += log2π
+    AE += m_γ * (V_y + m_y^2 - 2m_γ * m_y * m_θ'm_x + mul_trace(V_θ, V_x) + m_x'V_θ * m_x + m_θ' * (V_x + m_x * m_x') * m_θ)
+    AE /= 2
+    return AE
+end
```
Comment (albertpod) [inline src/nodes/softdot.jl:22]: You can safely use GammaDistributionFamily here.

I would suggest to remove multiline AE calculations and have smth like:

```julia
function average_energy(q_y::NormalDistributionsFamily, q_θ::NormalDistributionsFamily, q_x::NormalDistributionsFamily, q_γ::GammaDistributionsFamily)
    m_y, V_y = mean_cov(q_y)
    m_θ, V_θ = mean_cov(q_θ)
    m_x, V_x = mean_cov(q_x)
    m_γ = mean(q_γ)
    AE = -mean(log.(q_γ)) + log2π .+ m_γ .* (V_y + m_y.^2 .- 2m_γ * (m_θ'*m_x) .+ mul_trace(V_θ, V_x) .+ m_x' * V_θ * m_x .+ m_θ' * (V_x .+ m_x * m_x') * m_θ) ./ 2
    return AE
end
```

File: src/rules/softdot/gamma.jl [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Belief propagation: does not exist for softdot
```
Comment (albertpod) [inline src/rules/softdot/gamma.jl:1]: Good comment, although then be consistent and say which kind of variational rule is available, i.e., mean-field variational message

File: src/rules/softdot/x.jl [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Belief propagation: does not exist for softdot
+
+# Variational MP
+@rule softdot(:x, Marginalisation) (q_y::Any, q_θ::Any, q_γ::Any) = begin
+    my = mean(q_y)
+    mθ, Vθ = mean_cov(q_θ)
+    mγ = mean(q_γ)
+    Dx = mγ * (Vθ + mθ * mθ')
+    zx = mγ * mθ * my'
```
Comment (albertpod) [inline src/rules/softdot/x.jl:9]: unnecessary scalar transpose

**Review by albertpod — CHANGES_REQUESTED** [review]
Look good @abpolym! I've left just a few comments.

Don't forget the about formatter and double-check `average_energy` tests. They are failing.

**Review by albertpod — APPROVED** [review]

**Comment by albertpod** [comment]
@abpolym thanks for incorporating the comments. I approve.

**Comment by bvdmitri** [comment]
Great work @abpolym ! Thanks!

**Pull request merged**